### PR TITLE
feat: Enable API users to request sections with regex patterns

### PR DESCRIPTION
**Pull request by MthwRobinson**
### Summary

Adds an `m_section_regex` parameter to the `pipeline_api` function, making a `section_regex` file parameter available in the API. Enables users to request a section by passing in a regex pattern. Requested patterns are treated as raw strings. Enforces a 60 second timeout on regex execution to defend against regex denial of service attacks. A `TimeoutError` will result if the regex runs too long.

### Testing

The following API call should work after running `make run-app-dev`:

```
curl -X 'POST' \
  'http://localhost:8000/sec-filings/v0.0.1/section' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'file=@rgld-10-K-85535-000155837021011343.xbrl' \
  -F 'section_regex=risk factors' \
-F 'section=RISK_FACTORS' | jq -C . | less -R
```

The resulting JSON should have both `RISK_FACTORS` and `REGEX_0` keys.

### CODE REVIEW

File: pipeline-notebooks/pipeline-section.ipynb [hunk 1]
```diff
@@ -62,24 +62,7 @@
    "execution_count": null,
    "id": "be77e966",
    "metadata": {},
-   "outputs": [
```
Comment (MthwRobinson) [inline pipeline-notebooks/pipeline-section.ipynb:65]: @qued - Did this mess up your toggle switch stuff at all? I didn't see the `<script>` cell, but wasn't sure if I accidentally deleted it or something.

Comment (MthwRobinson) [reply pipeline-notebooks/pipeline-section.ipynb:65]: Nevermind, saw there's the separate PR open for that still

File: test_sec_filings/sec_filings/test_section_api.py [hunk 2]
```diff
@@ -65,3 +65,41 @@ def test_risk_narrative_api(form_type, section, tmpdir):
...
+    with open(filename, "w") as f:
+        f.write(sample_document)
+
+    # NOTE(robinson) - Reset the rate limit to avoid 429s in tests
+    app.state.limiter.reset()
+    client = TestClient(app)
+    response = client.post(
+        "sec-filings/v0.0.1/section",
+        files={"file": (filename, open(filename, "rb"), "text/plain")},
+        data={"section_regex": ["risk factors"]},
```
Comment (cragwolfe) [inline test_sec_filings/sec_filings/test_section_api.py:90]: looks good, but how about some more test regex's with special characters?

e.g. `'item\s1[AB]?\.'`, `^SECURITY AND EXCHANGE COMISSION FILING$`

File: prepline_sec_filings/api/section.py [hunk 3]
```diff
@@ -66,6 +106,11 @@ def pipeline_api(text, m_section=[]):
         results[section] = sec_document.get_section_narrative(
             section_string_to_enum[section]
         )
+    for i, section_regex in enumerate(m_section_regex):
+        regex_enum = get_regex_enum(section_regex)
+        with timeout(seconds=60):
```
Comment (cragwolfe) [inline prepline_sec_filings/api/section.py:111]: For a single regex call on a single narrative element, isn't anything over 1 second already pretty much egregious?

Comment (MthwRobinson) [reply prepline_sec_filings/api/section.py:111]: Yeah I can chop that down. I'll make it 5 seconds

File: test_sec_filings/sec_filings/test_section_api.py (hunk 2 above)
Comment (MthwRobinson) [reply test_sec_filings/sec_filings/test_section_api.py:92]: Added a second test with `^(?:prospectus )?summary$` to get some special characters in there

File: test_sec_filings/sec_filings/test_section_api.py [hunk 4]
```diff
@@ -65,3 +67,75 @@ def test_risk_narrative_api(form_type, section, tmpdir):
...
+        },
+        {
+            "text": "The business could be attacked by bears.",
+            "type": "NarrativeText",
+        },
+    ]
+
+
+@pytest.mark.parametrize(
+    "form_type, section",
```
[the remark this answers is not in the document]
Comment (cragwolfe) [reply test_sec_filings/sec_filings/test_section_api.py:111]: (same with prior test)

[the remark this answers is not in the document]
Comment (MthwRobinson) [reply test_sec_filings/sec_filings/test_section_api.py]: Nope, removed

File: prepline_sec_filings/api/section.py (hunk 3 above)
Comment (cragwolfe) [reply prepline_sec_filings/api/section.py]: For the record, my earlier comment "single regex call on a single narrative element" also isn't true, it's potentially for all elements. Still, 5 seconds sounds good.

**Comment by cragwolfe** [comment]
LGTM once i can figure out to pass special regex characters via the API. May be worth adding to the README as well.

**Comment by MthwRobinson** [comment]
This works via the API with special characters. I'll add it to the README

```
curl -X 'POST' \
  'http://localhost:8000/sec-filings/v0.0.1/section' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'file=@rgld-10-K-85535-000155837021011343.xbrl' \
  -F 'section_regex=^properties$'  | jq -C . | less -R
```

File: pipeline-notebooks/pipeline-section.ipynb [hunk 5]
```diff
@@ -1061,6 +866,18 @@
...
+   "codemirror_mode": {
+    "name": "ipython",
+    "version": 3
+   },
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.8.13"
```
Comment (qued) [inline pipeline-notebooks/pipeline-section.ipynb:880]: Should probably remove this metadata.

Comment (MthwRobinson) [reply pipeline-notebooks/pipeline-section.ipynb]: Removed

**Review by qued — APPROVED** [review]
LGTM, but you'll probably need https://github.com/Unstructured-IO/pipeline-sec-filings/pull/9 to go through to pass the CI checks.

**Comment by cragwolfe** [comment]
> -F 'file=@rgld-10-K-85535-000155837021011343.xbrl' \
>   -F 'section_regex=^properties$'  | jq -C . | less

how about with a backslash special character like \s? (maybe this is what was causing me issues)

**Comment by MthwRobinson** [comment]
@cragwolfe - I think I got the `\s` characters working, standby for the fix

**Comment by MthwRobinson** [comment]
@cragwolfe - This should work now

```
 curl -X 'POST' \
  'http://localhost:8000/sec-filings/v0.0.1/section' \
  -H 'accept: application/json' \
  -H 'Content-Type: multipart/form-data' \
  -F 'file=@rgld-10-K-85535-000155837021011343.xbrl' \
  -F "section_regex=^(\S+\W?)+$"
```

**Comment by cragwolfe** [comment]
> @cragwolfe - This should work now
> 
worth updating README?

**Review by cragwolfe — APPROVED** [review]
🎉

**Pull request merged**